src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,489 @@
+        /// This count includes the index block.
+        table_blocks_total: ?usize = null,
+
+        /// Invoked when the table index block and all content blocks have been written.
+        callback: *const fn (*RepairTable, RepairTableResult) void,
+        /// "next" belongs to the `faulty_tables` FIFO.
+        next: ?*RepairTable = null,
+    };
+
+    pub const Options = struct {
+        /// Lower-bound for the limit of concurrent request_block()'s available.
+        blocks_max: usize,
+        /// Maximum number of concurrent request_table()'s.
+        tables_max: usize,
+    };
+
+    options: Options,
+
+    /// Invariants:
+    /// - For every block address in faulty_blocks, ¬free_set.is_free(address).
+    faulty_blocks: FaultyBlocks,
+    /// Index within `faulty_blocks`, used to cycle through block-repair requests.
+    ///
+    /// Invariants:
+    /// - faulty_blocks.count() > 0 implies faulty_blocks_repair_index < faulty_blocks.count()
+    /// - faulty_blocks.count() = 0 implies faulty_blocks_repair_index = faulty_blocks.count()
+    faulty_blocks_repair_index: usize = 0,
+
+    /// Invariants:
+    /// - For every index address in faulty_tables: ¬free_set.is_free(address).
+    faulty_tables: FIFO(RepairTable) = .{ .name = "grid_repair_queue_tables" },
+
+    checkpointing: ?struct {
+        /// The number of faulty_blocks with state=aborting.
+        aborting: usize,
+    } = null,
+
+    /// Guard against new block/table repairs queueing within RepairTable callbacks.
+    canceling: bool = false,
+
+    pub fn init(allocator: std.mem.Allocator, options: Options) error{OutOfMemory}!GridRepairQueue {
+        var faulty_blocks = FaultyBlocks{};
+        errdefer faulty_blocks.deinit(allocator);
+
+        try faulty_blocks.ensureTotalCapacity(
+            allocator,
+            options.blocks_max + options.tables_max * lsm_table_content_blocks_max,
+        );
+
+        return GridRepairQueue{
+            .options = options,
+            .faulty_blocks = faulty_blocks,
+        };
+    }
+
+    pub fn deinit(queue: *GridRepairQueue, allocator: std.mem.Allocator) void {
+        assert(!queue.canceling);
+
+        queue.faulty_blocks.deinit(allocator);
+
+        queue.* = undefined;
+    }
+
+    /// When the queue wants more blocks than fit in a single request message, successive calls
+    /// to this function cycle through the pending BlockRequests.
+    pub fn block_requests(queue: *GridRepairQueue, requests: []vsr.BlockRequest) usize {
+        assert(!queue.canceling);
+
+        const faults_total = queue.faulty_blocks.count();
+        if (faults_total == 0) return 0;
+        assert(faults_total > queue.faulty_blocks_repair_index);
+
+        const fault_addresses = queue.faulty_blocks.entries.items(.key);
+        const fault_data = queue.faulty_blocks.entries.items(.value);
+
+        var requests_count: usize = 0;
+        var fault_offset: usize = 0;
+        while (requests_count < requests.len and
```

Comment:
Actually I'm going to keep the `while` loop -- we need `fault_offset` after the loop finishes to update the `queue.faulty_blocks_repair_index` cursor, and I want that update outside of the loop since it mutates the queue.